vmm/src/client.rs

```diff
@@ -0,0 +1,136 @@
+use common::message::{
+    serialize, ApplyMessage, ArcaHandle, CreateBlobMessage, CreateThunkMessage, CreateTreeMessage,
+    DropMessage, MessageParser, MessageSerializer, Messages, ReplyMessage, RunThunkMessage,
+};
+use common::refcnt::RefCnt;
+use common::ringbuffer::RingBuffer;
+use common::BuddyAllocator;
+extern crate alloc;
+use alloc::collections::VecDeque;
+use alloc::sync::Arc;
+use std::sync::RwLock;
+
+pub struct ArcaRef<'a> {
+    offset: Option<usize>,
+    client: Arc<RwLock<Client<'a>>>,
+}
+
+impl<'a> ArcaRef<'a> {
+    fn new(offset: usize, client: Arc<RwLock<Client<'a>>>) -> Self {
+        Self {
+            offset: Some(offset),
+            client,
+        }
+    }
+}
+
+impl Drop for ArcaRef<'_> {
+    fn drop(&mut self) {
+        match self.offset {
+            Some(o) => {
+                println!("Dropped");
+                self.client.write().unwrap().drop_handle(o);
+            }
+            None => {}
+        }
+    }
+}
+
+pub struct Client<'a> {
+    parser: MessageParser<'a>,
+    serializer: MessageSerializer<'a>,
+    pending_messages: VecDeque<(u8, Box<[u8]>)>,
+}
+
+impl<'a> Client<'a> {
```

Comment:
This is not idiomatic; ArcaRef should have additional newtypes (`BlobRef`, `TreeRef`, etc.) which expose methods that send the relevant messages for that type. The whole mandatory "buffer messages and send them all at once later and later get a reply" is weird; this struct should just provide an interface to send a bunch of messages immediately and block until all the replies arrive, and maybe later an async version which doesn't block, and the user can buffer themselves if they so desire.